### ISSUE TITLE
PermissionError: [Errno 13] Permission denied: '/downloads/.metube'

### ISSUE THREAD

**Issue by TDMatthews** [alexta69/metube#227]
I've been running the container a couple years and it just started erroring. I'm not sure if its a new version of the container or changes on the Synology (running in docker). I confirmed the UID/GID are set properly and it shows its running with those values in the log. I changed the permissions on .\downloads\* to 777 but still fails. Any ideas?

**Comment by alexta69** [comment]
Can you try downgrading to one of the previous versions (you can see the available tags here: https://hub.docker.com/r/alexta69/metube/tags) and see if it solves it? Just to understand whether it's a problem with the version, or with something else.

**Comment by TDMatthews** [comment]
I hadn’t updated in forever, which made it more challenging…

Going back through old versions, 22-06-19 is the first that fails. If I delete the .metube folder, it does recreate it but as root, and you will notice that its using 1000:1000 even though the proper IDs are specified in the environment section

Setting umask to 022
Creating download directory /downloads and state directory /downloads/.metube
Changing ownership of download and state directories to 1000:1000
Running MeTube as user 1000:1000
Traceback (most recent call last):
  File "app/main.py", line 72, in <module>
    dqueue = DownloadQueue(config, Notifier())
  File "/app/app/ytdl.py", line 189, in __init__
    self.queue = PersistentQueue(self.config.STATE_DIR + '/queue')
  File "/app/app/ytdl.py", line 143, in __init__
    os.mkdir(pdir)
PermissionError: [Errno 13] Permission denied: '/downloads/.metube'


Without changing anything else, if I edit the container, set the version to 2022-05-25, and recreate it, I get (and that is without deleting the .metube folder that 06-19 created but couldn’t use):

DEBUG:asyncio:Using selector: EpollSelector
INFO:ytdl:waiting for item to download
======== Running on http://0.0.0.0:8081 ========
(Press CTRL+C to quit)


For now I will leave it at 05-25



From: Alex ***@***.***>
Date: Tuesday, February 28, 2023 at 1:49 PM
To: alexta69/metube ***@***.***>
Cc: TDMatthews ***@***.***>, Author ***@***.***>
Subject: Re: [alexta69/metube] PermissionError: [Errno 13] Permission denied: '/downloads/.metube' (Issue #227)

Can you try downgrading to one of the previous versions (you can see the available tags here: https://hub.docker.com/r/alexta69/metube/tags) and see if it solves it? Just to understand whether it's a problem with the version, or with something else.

**Comment by alexta69** [comment]
You'll have a very old version of yt-dlp then, and it won't work with many video sites. It's really worth trying to solve the problem.
When you recreate the container with an empty downloads dir (remove the .metube that's there), it's failing to create it?

**Comment by TDMatthews** [comment]
No, all versions will create the .metube folder in downloads when it doesn't exist. Its just anything newer than 2022-05-25 can't access it.

Note that my downloads folder is never empty. Its downloading into an existing Synology folder thats mapped inside a Plex container. That never caused issues before though

**Comment by PikuZheng** [comment]
I firmly believe this is a problem with dsm, please tell me the version of dsm and the pid of your admin user？
In my env, the admin user is 1024:100
```
root@DS:~# cat /etc/passwd
ActiveBackup:x:192868:192868::/var/packages/ActiveBackup/target:/sbin/nologin
admin:x:1024:100:System default user:/var/services/homes/admin:/bin/sh
anonymous:x:21:21::/nonexist:/usr/bin/nologin
```

**Comment by TDMatthews** [comment]
I think you are right, I'm just not sure on the fix.

/volume2/Entertainment/YouTube$ id
uid=1026(tmatthew) gid=100(users) groups=100(users),101(administrators),65537(docker)

I recreated the container with latest, deleted .metube and ran as 1026 100. It creates the .metube folder as my tmatthew user but then exits.

Setting umask to 022
Creating download directory /downloads and state directory /downloads/.metube
Changing ownership of download and state directories to **1026:100**
Running MeTube as user **1026:100**
Traceback (most recent call last):
  File "app/main.py", line 85, in <module>
    dqueue = DownloadQueue(config, Notifier())
  File "/app/app/ytdl.py", line 195, in __init__
    self.queue = PersistentQueue(self.config.STATE_DIR + '/queue')
  File "/app/app/ytdl.py", line 149, in __init__
    os.mkdir(pdir)
PermissionError: [Errno 13] Permission denied: '/downloads/.metube'

I saw another user suggest this. If I change nothing else (don't even delete .metube) and redeploy as uid=0 gid=0 (root) it works no problem:

Setting umask to 022
Creating download directory /downloads and state directory /downloads/.metube
Warning: it is not recommended to run as root user, please check your setting of the UID environment variable
Changing ownership of download and state directories to 0:0
Running MeTube as user 0:0
DEBUG:asyncio:Using selector: EpollSelector
INFO:ytdl:waiting for item to download


I don't really like that from a best practices point of view, but I guess the risk is low. It's only mapped to the one Plex folder that just has YT downloads in it, so even going rogue it can't do much damage, I just would like to solve the underlying issue

**Comment by PikuZheng** [comment]
I can reproduce this problem. . . I will update once there is any progress.
```
root@DS:~# docker create --name=metube -v /volume2/video:/downloads -p 8081:8081/tcp alexta69/metube
Unable to find image 'alexta69/metube:latest' locally
latest: Pulling from alexta69/metube
63b65145d645: Pull complete
781ed.....

root@DS:~# docker start metube
metube
root@DS:~# docker logs metube
Setting umask to 022
Creating download directory /downloads and state directory /downloads/.metube
Changing ownership of download and state directories to 1000:1000
Running MeTube as user 1000:1000
Traceback (most recent call last):
  File "app/main.py", line 85, in <module>
    dqueue = DownloadQueue(config, Notifier())
  File "/app/app/ytdl.py", line 195, in __init__
    self.queue = PersistentQueue(self.config.STATE_DIR + '/queue')
  File "/app/app/ytdl.py", line 149, in __init__
    os.mkdir(pdir)
PermissionError: [Errno 13] Permission denied: '/downloads/.metube'
```

**Comment by PikuZheng** [comment]
In DSM, add "read and write" permissions for "Owner" on the downloaded target folder. No need to modify container settings.
![image](https://user-images.githubusercontent.com/12579522/222034217-e50df66f-2f85-4da0-bdfa-7426429adf95.png)

Details if you want to know: 
The default shared folder permission of synology is 000. Even if umask=022, the new folder permission is still (000-022=)000.
Although the folder owner is 1000:1000, user 1000 cannot read and write the folder with permission 000 during su-exec (only user 0 can read and write).
Add "read and write" permissions for "Owner" will change permisson to 700, so the owner (1000) can r/w correctly.

**Comment by alexta69** [comment]
Closing (not a MeTube issue, if you find out otherwise please reopen)

**Comment by Gibado** [comment]
This was happening to me as well.  I finally found a solution to make it work so I'll leave what I did here in case someone else has this issue.

After starting the container it would have permission issues while attempting to access `/downloads/.metube`.  I stopped the container manually set the permissions of `.metube` to 777 
```
sudo chmod -R 777 metube-directory/.metube
```
Next I started the container back up and I noticed more permission issues, but this time it was for `/downloads/.metube/completed`, `/downloads/.metube/pending`, and `/downloads/.metube/queue`.  These files all had permissions set to `000`.  

Next I repeated the last step of stopping the container, updating the permissions of all directories/files in the metube directory, and starting the container again.  Now it's happy.

No idea why it couldn't update the permissions when it was the owner, but I'm glad it's working again.
